### PR TITLE
NEXTEUROPA-12288: Enable Composer support for sub-sites

### PR DESCRIPTION
**Pull request by ademarco**

### CODE REVIEW

**Comment by ademarco** [comment]
\cc #15

File: build.package.xml [hunk 1]
```diff
@@ -174,10 +174,13 @@
...
-        <!-- See https://webgate.ec.europa.eu/CITnet/jira/browse/NEXTEUROPA-11706 for more information. -->
-        <symlink link="bin/behat" target="${platform.build.dir}/vendor/bin/behat" overwrite="true"/>
-        <symlink link="${platform.build.dir}/tests" target="${project.basedir}/tests" overwrite="true"/>
+    </target>
+
+    <!-- Symlink the Behat bin and test directory in './sites/all'. -->
+    <target name="link-behat">
+        <echo msg="Symlink the Behat bin and test directory in './sites/all'." />
+        <symlink link="bin/behat" target="${platform.build.subsite.dir}/vendor/bin/behat" overwrite="true"/>
+        <symlink link="${platform.build.subsite.dir}/tests" target="${project.basedir}/tests" overwrite="true"/>
```
Comment (pfrenssen) [inline build.package.xml]: I'm wondering why we added this originally, there is little value of having this symlink inside the Drupal installation. Maybe we were thinking to allow to run these tests on staging or acceptance? I don't remember but it seems useless. This is out of scope for this issue though, so let's ignore it for now.

File: build.package.xml [hunk 2]
```diff
@@ -174,10 +174,13 @@
         <!-- Symlink composer configuration. -->
         <symlink link="${platform.build.subsite.composer.json}" target="${subsite.resources.composer.json}" />
         <symlink link="${platform.build.subsite.composer.lock}" target="${subsite.resources.composer.lock}" />
-        <!-- Symlink Behat binary and test directory from platform build. -->
-        <!-- See https://webgate.ec.europa.eu/CITnet/jira/browse/NEXTEUROPA-11706 for more information. -->
-        <symlink link="bin/behat" target="${platform.build.dir}/vendor/bin/behat" overwrite="true"/>
-        <symlink link="${platform.build.dir}/tests" target="${project.basedir}/tests" overwrite="true"/>
+    </target>
+
+    <!-- Symlink the Behat bin and test directory in './sites/all'. -->
```
Comment (pfrenssen) [inline build.package.xml]: I wouldn't mention 'sites/all'. This is the default location but since it is configurable it is not guaranteed that this is the actual location. Just say 'in the subsite folder'' or something.

File: build.package.xml [hunk 3]
```diff
@@ -214,24 +217,26 @@
...
         </composer>
     </target>
 
-    <!-- Install Composer development dependencies for the platform. -->
-    <!-- Note: this is a temporary workaround until we get Composer to work properly with sub-sites. -->
-    <!-- See https://webgate.ec.europa.eu/CITnet/jira/browse/NEXTEUROPA-11706 for more information. -->
-    <target name="install-platform-dev-dependencies">
+    <!-- Install Composer dist dependencies for the subsite. -->
+    <target name="install-dist-dependencies">
+        <echo msg="Run 'composer install --no-dev' in './build'." />
```
Comment (pfrenssen) [inline build.package.xml]: Same here, the build location is configurable, avoid mentioning the absolute path.

File: build.package.xml [hunk 4]
```diff
@@ -214,24 +217,26 @@
...
+        <echo msg="Run 'composer install' in platform root." />
         <composer command="install" composer="${composer.bin}">
             <arg value="--working-dir=${project.basedir}" />
         </composer>
     </target>
 
-    <!-- Install Composer dependencies for the subsite. -->
+    <!-- Install Composer dev dependencies for the subsite. -->
     <target name="install-dev-dependencies">
+        <echo msg="Run 'composer install' in './sites/all' for sub-site development purposes." />
```
Comment (pfrenssen) [inline build.package.xml]: `s/\'\.\/sites\/all/the subsite folder/`

File: build.package.xml [hunk 5]
```diff
@@ -214,24 +217,26 @@
...
             <arg value="--working-dir=${platform.build.subsite.dir}" />
         </composer>
     </target>
 
-    <!-- Install Composer development dependencies for the platform. -->
-    <!-- Note: this is a temporary workaround until we get Composer to work properly with sub-sites. -->
-    <!-- See https://webgate.ec.europa.eu/CITnet/jira/browse/NEXTEUROPA-11706 for more information. -->
-    <target name="install-platform-dev-dependencies">
+    <!-- Install Composer dist dependencies for the subsite. -->
+    <target name="install-dist-dependencies">
```
Comment (pfrenssen) [inline build.package.xml]: Thanks for implementing this! We were missing this functionality.

File: build.package.xml [hunk 6]
```diff
@@ -313,13 +318,28 @@
...
+            delete-dev,
+            make-dev,
+            link-dev-resources,
+            install-dev-dependencies,
+            link-behat,
+            setup-behat,
+            update-htaccess,
+            setup-php-codesniffer,
+            download-development-modules"
     />
```
Comment (pfrenssen) [inline build.package.xml]: :+1:

File: resources/composer.json [hunk 7]
```diff
@@ -1,9 +1,14 @@
...
-        "psr-4": {
-            "NextEuropa\\": "src"
-        }
+  "require": {
+    "guzzlehttp/guzzle": "~6.0"
+  },
+  "require-dev": {
+    "behat/behat": "~3.1.0@rc",
+    "drupal/drupal-extension": "~3.1.0"
+  },
```
Comment (pfrenssen) [inline resources/composer.json]: Is this added as an example? I'm not too happy with this, the 80% case is that a site will not need any of these dependencies. If we add this they will be installed on every dev's machine from every project. These are large projects with a lot of dependencies of their own.

I'm actually not sure about the use case for having a dev dependency inside the subsite. Especially not behat, this is standalone and is already present in the project root, we don't need a duplicate of this.

Guzzle is a good example, but I'd rather have this in the README than for real in the {{composer.json}}.

Comment (ademarco) [inline resources/composer.json]: Guzzle is actually a Platform requirement, since the introduction of the NextEuropa Integration module.

I figured that the sub-site composer.json will need to be manually maintained when dependencies will be updated at platform level: it will be then responsibility of the sub-site maintainer to re-align manually his own composer.json with the platform one.

About Behat we figured that, conceptually, Behat is not really a dependency of the "builder" composer.json (i.e. the root composer.json) but it's a dev dependency of the Drupal site and, as such, its place should be `require-dev` of the site composer.json.

In fact, in a distant future, we could theoretically stop to build platform and sub-sites with Phing and this should not concern Behat, which will always be a dev dependency of the platform and of its sub-sites, regardless how we build them.

We will take care of updating documentation in order to reflect the changes outlined above.

File: build.package.xml (hunk 2 above)
Comment (ademarco) [inline build.package.xml]: Ok I'll change that, and the other occurrences as well, thanks!

File: resources/composer.json (hunk 7 above)
Comment (pfrenssen) [inline resources/composer.json]: OK makes sense, thanks!

In the meanwhile I've grown used to structuring a project according to how it is done in https://github.com/drupal-composer/drupal-project - that has a single `composer.json` that has all the dependencies of the project. It does not make a distinction between the 'build system' and the 'drupal site'. It simply contains everything in a single file.

For subsite starterkit it is not a good idea to change this at this time. We are still using drush make for most of the dependencies.

Comment (ademarco) [inline resources/composer.json]: Yeah, moving to a completely Composer-based workflow would solve many issues indeed.

**Comment by verbruggenalex** [comment]
This change is really welcome since we are using composer to grab packages for sub-platforms like nems and reps. At the moment we are allowing the sites to hack the main composer.json: https://github.com/ec-europa/nems-platform-reference/blob/master/readme.md We want to switch to the subsite composer.json as soon as possible.

**Comment by pfrenssen** [comment]
@ademarco I have made you administrator of this repository so you can merge issues. Then you don't need to wait for me to become available. Just follow standard procedure: make sure a full QA is done by a colleague before merging.

File: resources/composer.json (hunk 7 above)
Comment (pfrenssen) [inline resources/composer.json]: That will probably not be on the current road map, more chance when we start porting the platform to D8 :)
